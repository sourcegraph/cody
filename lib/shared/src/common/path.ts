import type { URI } from 'vscode-uri'

import { isWindows as _isWindows, isMacOS } from './platform'

interface PathFunctions {
    /**
     * All but the last element of path, or "." if that would be the empty path.
     */
    dirname: (path: string) => string

    /**
     * The last element of path, or "" if path is empty.
     * @param path the path to operate on
     * @param suffix optional suffix to remove
     */
    basename: (path: string, suffix?: string) => string

    /** The extension of path, including the last '.'. */
    extname: (path: string) => string

    /**
     * The relative path from {@link from} to {@link to}.
     */
    relative: (from: string, to: string) => string

    resolve: (...paths: string[]) => string

    /** Path separator. */
    separator: PathSeparator
}

/** For file system paths on Windows ('\' separators, drive letters, case-insensitive). */
export const windowsFilePaths: PathFunctions = pathFunctions(true, '\\', false)

/** For POSIX file system paths ('/' separators, case-sensitive). */
export const posixFilePaths: PathFunctions = pathFunctions(false, '/', true)

/**
 * Get the {@link PathFunctions} to use for the given URI's path ('/' separators, drive letters/case-sensitivity depend on `isWindows`).
 */
export function pathFunctionsForURI(uri: URI, isWindows = _isWindows()): PathFunctions {
    // URIs are always forward slashes even on Windows.
    const sep = '/'
    return uri.scheme === 'file' && isWindows
        ? // Like windowsFilePaths but with forward slashes.
          pathFunctions(true, sep, false)
        : posixFilePaths
}

export function defaultPathFunctions(): PathFunctions {
    const isWindows = _isWindows()
    return pathFunctions(isWindows, isWindows ? '\\' : '/', isMacOS())
}

// I don't like reimplementing this here, but it's the best option because: (1) using Node's `path`
// module requires us to configure the bundler right or risk a subtle bug (and path-browserify
// doesn't have `path/win32` support, so we'd need multiple underlying packages); and (2)
// `vscode-uri` is hard to test because it has a global constant for the current platform set at
// init time.
//
// note: case-insensitive for Windows is not strictly sound, it's possible to have case-sensitive
// volumes on Windows (and case-insensitive on other platforms), however it's impossible to tell from
// a URI/path alone whether that's the case and VS Code already assumes based on platform, so things
// are unfortunately already broken if you don't use the "defaults".
// - https://github.com/microsoft/vscode/issues/94307
// - https://github.com/microsoft/vscode/issues/177925
function pathFunctions(isWindows: boolean, sep: '\\' | '/', caseSensitive: boolean): PathFunctions {
    const f: PathFunctions = {
        dirname(path: string): string {
            if (path === '') {
                return '.'
            }
            if (isWindows && isDriveLetter(path)) {
                return path
            }
            if (path.endsWith(sep)) {
                path = path.slice(0, -1)
            }
            if (isWindows && isDriveLetter(path)) {
                return path + sep
            }
            if (path === '') {
                return sep
            }
            const i = path.lastIndexOf(sep)
            if (i === -1) {
                return '.'
            }
            if (i === 0) {
                return sep
            }
            path = path.slice(0, i)
            if (isWindows && isDriveLetter(path)) {
                return path + sep
            }
            return path
        },
        basename(path: string, suffix?: string): string {
            if (path.endsWith(sep)) {
                path = path.slice(0, -1)
            }
            if (isWindows && isDriveLetter(path)) {
                return ''
            }
            path = path.split(sep).at(-1) ?? ''
            if (suffix && path.endsWith(suffix)) {
                path = path.slice(0, -suffix.length)
            }
            return path
        },
        extname(path: string): string {
            const basename = f.basename(path)
            const i = basename.lastIndexOf('.')
            if (i === 0 || i === -1) {
                return ''
            }
            return basename.slice(i)
        },
        relative(from, to) {
            // For relative paths, output separator is always based on platform, even
            // if the input (sep) is forward slash/URI.
            return relative(from, to, sep, isWindows ? '\\' : '/', caseSensitive)
        },
        resolve(...paths) {
            return paths.join(sep)
        },
        separator: sep,
    }
    return f
}

function isDriveLetter(path: string): boolean {
    return /^[A-Za-z]:$/.test(path)
}

/** The relative path from {@link from} to {@link to}. */
function relative(
    from: string,
    to: string,
    inputSeparator: PathSeparator,
    outputSeparator: PathSeparator,
    caseSensitive: boolean
): string {
    function equalSegments(a: string, b: string) {
        return caseSensitive ? a === b : a.toLowerCase() === b.toLowerCase()
    }

    // Normalize extra slashes.
    if (inputSeparator === '/') {
        from = from.replaceAll(/\/{2,}/g, '/')
        to = to.replaceAll(/\/{2,}/g, '/')
    } else {
        from = from.replaceAll(/\\{2,}/g, '\\')
        to = to.replaceAll(/\\{2,}/g, '\\')
    }

    // Trim trailing slashes.
    if (from !== inputSeparator && from.endsWith(inputSeparator)) {
        from = from.slice(0, -1)
    }
    if (to !== inputSeparator && to.endsWith(inputSeparator)) {
        to = to.slice(0, -1)
    }

    if (equalSegments(from, to)) {
        return ''
    }

    // From a relative to an absolute path, the absolute path dominates.
    if (!from.startsWith(inputSeparator) && to.startsWith(inputSeparator)) {
        return to
    }

    const fromParts = from === inputSeparator ? [''] : from.split(inputSeparator)
    const toParts = to === inputSeparator ? [''] : to.split(inputSeparator)

    // Find the common root.
    let commonLength = 0
    while (
        commonLength < fromParts.length &&
        commonLength < toParts.length &&
        equalSegments(fromParts[commonLength], toParts[commonLength])
    ) {
        commonLength++
    }

    // Calculate the number of directory traversals needed.
    const traversals = fromParts.length - commonLength

    // Build the relative path.
    const relativePath: string[] = []
    for (let i = 0; i < traversals; i++) {
        relativePath.push('..')
    }

    // Add the non-common path parts from the 'to' path.
    relativePath.push(...toParts.slice(commonLength))

    return relativePath.join(outputSeparator)
}

type PathSeparator = '\\' | '/'
