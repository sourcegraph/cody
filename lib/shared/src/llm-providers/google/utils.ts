import type { Content, InlineDataPart, Part } from '@google/generative-ai'
import type { Message } from '../..'
import { getMessageImageUrl } from '../completions-converter'

/**
 * Constructs the messages array for the Gemini API, including handling InlineDataPart for media.
 */
export async function constructGeminiChatMessages(messages: Message[]): Promise<Content[]> {
    const contents: Content[] = []

    // Map speaker types to Gemini API roles
    const roleMap: Record<string, 'user' | 'model' | 'system'> = {
        human: 'user',
        assistant: 'model',
        system: 'system',
    }

    for (const message of messages) {
        const role = roleMap[message.speaker] || 'user'
        const parts: Part[] = []

        // Skip if this would create consecutive messages from the same role
        const lastContent = contents[contents.length - 1]
        if (lastContent?.role === role) {
            continue
        }

        // Process message content parts
        if (message.content?.length) {
            for (const part of message.content) {
                if (part.type === 'text' && part.text?.length) {
                    parts.push({ text: part.text })
                }
                const { data, mimeType } = getMessageImageUrl(part)
                if (data && mimeType) {
                    parts.push({
                        inlineData: { mimeType, data: data.replace(/data:[^;]+;base64,/, '') },
                    } satisfies InlineDataPart)
                }
            }
        }

        // Add message text if present
        if (message.text?.length) {
            parts.push({ text: message.text.toString() })
        }

        // Add content if there are parts
        if (parts.length > 0) {
            contents.push({ role, parts })
        }
    }

    // Remove trailing model message if present
    if (contents.length > 0 && contents[contents.length - 1].role === 'model') {
        contents.pop()
    }

    return contents
}
