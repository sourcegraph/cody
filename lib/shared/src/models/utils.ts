import type { ClientConfiguration } from '../configuration'
import { logDebug } from '../logger'
import { ANSWER_TOKENS } from '../prompt/constants'
import type { CodyLLMSiteConfiguration } from '../sourcegraph-api/graphql/client'
import {
    CHAT_INPUT_TOKEN_BUDGET,
    CHAT_OUTPUT_TOKEN_BUDGET,
    EXTENDED_CHAT_INPUT_TOKEN_BUDGET,
    EXTENDED_USER_CONTEXT_TOKEN_BUDGET,
} from '../token/constants'
import type { Model } from './model'
import type { ModelRef, ModelRefStr } from './modelsService'
import { ModelTag } from './tags'
import type { ModelContextWindow } from './types'

export function getProviderName(name: string): string {
    const providerName = name.toLowerCase()
    switch (providerName) {
        case 'anthropic':
            return 'Anthropic'
        case 'openai':
            return 'OpenAI'
        case 'ollama':
            return 'Ollama'
        case 'google':
            return 'Google'
        default:
            return providerName
    }
}

/**
 * Gets the provider and title from a model ID string.
 */
export function getModelInfo(modelID: string): {
    provider: string
    title: string
} {
    const [providerID, ...rest] = modelID.split('/')
    const provider = getProviderName(providerID)
    const title = (rest.at(-1) || '').replace(/-/g, ' ')
    return { provider, title }
}

export function isCodyProModel(model: Model): boolean {
    return modelHasTag(model, ModelTag.Pro)
}

export function isWaitlistModel(model: Model): boolean {
    return modelHasTag(model, ModelTag.Waitlist) || modelHasTag(model, ModelTag.OnWaitlist)
}

export function isCustomModel(model: Model): boolean {
    return (
        modelHasTag(model, ModelTag.Local) ||
        modelHasTag(model, ModelTag.Dev) ||
        modelHasTag(model, ModelTag.BYOK)
    )
}

function modelHasTag(model: Model, modelTag: ModelTag): boolean {
    return model.tags.includes(modelTag)
}

export function toModelRefStr(modelRef: ModelRef): ModelRefStr {
    const { providerId, apiVersionId, modelId } = modelRef
    return `${providerId}::${apiVersionId}::${modelId}`
}

/**
 * Get the context window for the given chat model and configuration overwrites in an enterprise environment.
 *
 * @param chatModel - The chat model to get the context window for.
 * @param configOverwrites - The configuration overwrites to apply.
 * @returns The context window for the given chat model and configuration overwrites.
 */
export function getEnterpriseContextWindow(
    chatModel: string,
    configOverwrites: CodyLLMSiteConfiguration,
    configuration: Pick<ClientConfiguration, 'providerLimitPrompt'>
): ModelContextWindow {
    const { chatModelMaxTokens, smartContextWindow } = configOverwrites
    // Starts with the default context window.
    let contextWindow: ModelContextWindow = {
        input: chatModelMaxTokens ?? CHAT_INPUT_TOKEN_BUDGET,
        output: getEnterpriseOutputLimit(chatModel),
    }

    // Use extended context window for models that support smart context when enabled.
    if (smartContextWindow && isModelWithExtendedContextWindowSupport(chatModel)) {
        contextWindow = {
            input: EXTENDED_CHAT_INPUT_TOKEN_BUDGET,
            output: CHAT_OUTPUT_TOKEN_BUDGET,
            context: { user: EXTENDED_USER_CONTEXT_TOKEN_BUDGET },
        }
    }

    return applyLocalTokenLimitOverwrite(configuration, chatModel, contextWindow)
}

/**
 * Applies a local token limit overwrite to the given context window if configured.
 * If the configured limit is lower than the default, it will be applied to the input of the context window.
 * If the configured limit is invalid, an error will be logged.
 *
 * @param chatModel The chat model for which the token limit is being applied.
 * @param contextWindow The context window to apply the token limit overwrite to.
 * @returns The updated context window with the token limit overwrite applied, or the original context window if no valid overwrite is configured.
 */
function applyLocalTokenLimitOverwrite(
    { providerLimitPrompt }: Pick<ClientConfiguration, 'providerLimitPrompt'>,
    chatModel: string,
    contextWindow: ModelContextWindow
): ModelContextWindow {
    if (providerLimitPrompt && providerLimitPrompt <= contextWindow.input) {
        return { ...contextWindow, input: providerLimitPrompt }
    }

    if (providerLimitPrompt) {
        logDebug(
            'getEnterpriseContextWindow',
            `Invalid token limit configured for ${chatModel}`,
            providerLimitPrompt
        )
    }

    return contextWindow
}

/**
 * Returns true if the given chat model supports extended context windows.
 *
 * @param chatModel - The name of the chat model.
 * @returns True if the chat model supports extended context windows, false otherwise.
 */
const modelWithExpandedWindowSubStrings = [
    'opus',
    'sonnet', // Covers claude 3 sonnet & 3.5 sonnet
    'gemini-1.5',
    'gpt-4o',
    'gpt-4-turbo',
]
function isModelWithExtendedContextWindowSupport(chatModel: string): boolean {
    return modelWithExpandedWindowSubStrings.some(keyword => chatModel.toLowerCase().includes(keyword))
}

// TODO: Currently all enterprise models have a max output limit of
// 1000. We need to support configuring the maximum output limit at an
// instance level. This will allow us to increase this limit whilst
// still supporting models with a lower output limit.
//
// To avoid Enterprise instances being stuck with low token counts, we
// will detect our recommended Cody Gateway models to use a higher limit.
//
// See: https://github.com/sourcegrcaph/cody/issues/3648#issuecomment-2056954101
// See: https://github.com/sourcegraph/cody/pull/4203
function getEnterpriseOutputLimit(model?: string) {
    if (model && isModelWithExtendedContextWindowSupport(model)) {
        return CHAT_OUTPUT_TOKEN_BUDGET
    }
    return ANSWER_TOKENS
}
