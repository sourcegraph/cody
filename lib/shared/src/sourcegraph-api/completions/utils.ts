import { type SerializedChatMessage, contextFiltersProvider } from '../..'
import { serverSupportsPromptCaching } from '../clientConfig'
import type { CompletionParameters, Message, SerializedCompletionParameters } from './types'

/**
 * Serializes the completion parameters by converting the message text to a filtered string using the provided context filters.
 *
 * @param params - The completion parameters to serialize.
 * @returns A serialized version of the completion parameters with the message text filtered.
 */
export async function getSerializedParams(
    params: CompletionParameters
): Promise<SerializedCompletionParameters> {
    return {
        ...params,
        messages: await serializePrompts(params.messages, params.model),
    }
}

async function serializePrompts(
    messages: Message[],
    modelID?: string
): Promise<SerializedChatMessage[]> {
    // NOTE: Some models do not support empty assistant message at the end.
    if (modelID?.startsWith('google/') && messages.at(-1)?.speaker === 'assistant') {
        messages.pop()
    }

    return Promise.all(
        messages.map(async m => {
            const text = await m.text?.toFilteredString(contextFiltersProvider)
            if (serverSupportsPromptCaching() && m.cacheEnabled) {
                return {
                    speaker: m.speaker,
                    content: [
                        {
                            ...m,
                            type: 'text',
                            text: text ?? '',
                            cache_control: { type: 'ephemeral' },
                        },
                    ],
                }
            }
            return {
                ...m,
                text: text,
            }
        })
    )
}
