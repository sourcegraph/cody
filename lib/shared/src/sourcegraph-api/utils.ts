// Converts the prefix of buf to a UTF8 string. If buf terminates in the middle
// of a character, returns the remaining bytes of the partial character in a
// new buffer. Note! This assumes that the prefix of buf *is* valid UTF8--it
// only examines the bytes of the last character in the buffer and assumes it
// will find an initial byte before the start of the buffer.
export function toPartialUtf8String(buf: Buffer): { str: string; buf: Buffer } {
    if (buf.length === 0) {
        return { str: '', buf: Buffer.of() }
    }
    let lastValidByteOffsetExclusive = buf.length
    if ((buf[lastValidByteOffsetExclusive - 1] & 0x80) !== 0) {
        // Multi-byte character. Count additional trailing bytes. UTF8 trailing
        // bytes have the bit pattern 10??_????.
        let numBytes = 1
        while ((buf[lastValidByteOffsetExclusive - numBytes] & 0xc0) === 0x80) {
            numBytes++
        }
        // Scrutinize the initial byte to see if the encoding is complete.
        // Characters of N bytes encode the length in the first character.
        // The high order N bits are set, and the next bit is clear. For
        // example a 4-byte character starts with 1111_0???.
        const byte = buf[lastValidByteOffsetExclusive - numBytes]
        const mask = 0xff ^ ((1 << (7 - numBytes)) - 1)
        const value = numBytes === 6 ? 0xfc : mask ^ (1 << (7 - numBytes))
        if ((byte & mask) !== value) {
            // The trailing bytes are incomplete; don't decode them.
            lastValidByteOffsetExclusive -= numBytes
        }
    }
    return {
        str: buf.slice(0, lastValidByteOffsetExclusive).toString('utf8'),
        buf: Buffer.from(buf.slice(lastValidByteOffsetExclusive)),
    }
}
