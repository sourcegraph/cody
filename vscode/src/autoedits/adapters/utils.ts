import { type Message, type PromptString, charsToTokens } from '@sourcegraph/cody-shared'

export interface FireworksCompatibleRequestParams {
    stream: boolean
    model: string
    temperature: number
    max_tokens: number
    response_format: {
        type: string
    }
    prediction: {
        type: string
        content: string
    }
    rewrite_speculation?: boolean
    user?: string
}

export function getMaxOutputTokensForAutoedits(codeToRewrite: string): number {
    const MAX_NEW_GENERATED_TOKENS = 512
    const codeToRewriteTokens = charsToTokens(codeToRewrite.length)
    return codeToRewriteTokens + MAX_NEW_GENERATED_TOKENS
}

export function getOpenaiCompatibleChatPrompt(param: {
    systemMessage?: PromptString
    userMessage: PromptString
}): { role: string; content: PromptString }[] {
    const prompt = []
    if (param.systemMessage) {
        prompt.push({ role: 'system', content: param.systemMessage })
    }
    prompt.push({ role: 'user', content: param.userMessage })
    return prompt
}

export function getSourcegraphCompatibleChatPrompt(param: {
    systemMessage: PromptString | undefined
    userMessage: PromptString
}): Message[] {
    const prompt: Message[] = []
    if (param.systemMessage) {
        prompt.push({ speaker: 'system', text: param.systemMessage })
    }
    prompt.push({ speaker: 'human', text: param.userMessage })
    return prompt
}

export async function getModelResponse(
    url: string,
    body: string,
    apiKey: string,
    customHeaders: Record<string, string> = {}
): Promise<any> {
    const response = await fetch(url, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            Authorization: `Bearer ${apiKey}`,
            ...customHeaders,
        },
        body: body,
    })
    if (response.status !== 200) {
        const errorText = await response.text()
        throw new Error(`HTTP error! status: ${response.status}, message: ${errorText}`)
    }
    const data = await response.json()
    return data
}
