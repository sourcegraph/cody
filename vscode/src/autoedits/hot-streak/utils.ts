import * as vscode from 'vscode'
import { LONG_SUGGESTION_USER_CURSOR_MARKER } from '../prompt/constants'

export function trimPredictionToLastFullLine(prediction: string): string {
    if (!prediction) {
        return prediction
    }

    // If the prediction ends with a newline, it's already complete
    if (prediction.endsWith('\n')) {
        return prediction
    }

    const lastNewlineIndex = prediction.lastIndexOf('\n')
    if (lastNewlineIndex === -1) {
        // If there's no newline, we can't trim to a complete line
        return ''
    }

    // Return everything up to and including the last newline
    return prediction.substring(0, lastNewlineIndex + 1)
}

export function postProcessCompletion(prediction: string): string {
    const cursorMarker = LONG_SUGGESTION_USER_CURSOR_MARKER.toString()
    return prediction.replace(cursorMarker, '')
}

export function isHotStreakEnabledInSettings() {
    return vscode.workspace
        .getConfiguration()
        .get<boolean>('cody.experimental.autoedit.use-hot-streak', false)
}
