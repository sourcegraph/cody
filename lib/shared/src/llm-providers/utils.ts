import type { CompletionsModelConfig } from '.'
import { modelsService } from '../models/modelsService'

export function getCompletionsModelConfig(modelID: string): CompletionsModelConfig | undefined {
    const provider = modelsService.getModelByID(modelID)
    if (!provider) {
        return undefined
    }

    const { id, clientSideConfig = {} } = provider
    const { apiKey = '', apiEndpoint, options = {} } = clientSideConfig

    const prefix = provider.provider.toLowerCase() + '/'
    const model = id.startsWith(prefix) ? id.substring(prefix.length) : id
    const stream = Boolean(options?.stream ?? true)

    const { stream: _, ...restOptions } = options || {}

    return {
        model,
        key: apiKey,
        endpoint: apiEndpoint,
        stream,
        options: restOptions,
    }
}
