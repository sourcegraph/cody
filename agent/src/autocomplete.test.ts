import path from 'node:path'
import { afterAll, beforeAll, describe, expect, it } from 'vitest'
import { sleep } from '../../vscode/src/completions/utils'
import { TESTING_CREDENTIALS } from '../../vscode/src/testutils/testing-credentials'
import { TestClient } from './TestClient'
import { TestWorkspace } from './TestWorkspace'

describe('Autocomplete', () => {
    const workspace = new TestWorkspace(path.join(__dirname, '__tests__', 'autocomplete'))
    const client = TestClient.create({
        workspaceRootUri: workspace.rootUri,
        name: 'autocomplete',
        credentials: TESTING_CREDENTIALS.enterprise,
    })

    beforeAll(async () => {
        await workspace.beforeAll()
        await client.beforeAll()
    })

    afterAll(async () => {
        await workspace.afterAll()
        await client.afterAll()
    })

    it('autocomplete/execute (non-empty result)', async () => {
        const uri = workspace.file('src', 'sum.ts')
        await client.openFile(uri)

        // Set a small visibility delay for testing purposes.
        const visibilityDelay = 100
        await client.request('testing/autocomplete/setCompletionVisibilityDelay', {
            delay: visibilityDelay,
        })

        // Generate completions for the current cursor position.
        const completions = await client.request('autocomplete/execute', {
            uri: uri.toString(),
            position: { line: 1, character: 4 },
            triggerKind: 'Automatic',
        })
        const completionID = completions.items[0].id

        // Tell completion provider that the completion was shown to the user.
        client.notify('autocomplete/completionSuggested', { completionID })

        // Wait for only half of the time required to mark the completion as read.
        await sleep(visibilityDelay / 2)

        // Change the cursor position discarding the current completion.
        // Since the completion was visible only `visibilityDelay / 2` it should not be marked as read.
        await client.changeFile(uri, {
            selection: {
                start: { line: 0, character: 0 },
                end: { line: 0, character: 0 },
            },
        })

        // Now the completion visibility timeout is elapsed but because we discarded it, it should not be
        // marked as read because we moved the cursor on the previous step and discarded the completion.
        await client.request('testing/autocomplete/awaitPendingVisibilityTimeout', null)

        // Get the analytics event for the completion to assert the read and accepted state.
        const completionEvent = await client.request('testing/autocomplete/completionEvent', {
            completionID,
        })
        expect(completionEvent?.read).toBe(false)
        expect(completionEvent?.acceptedAt).toBeFalsy()

        const texts = completions.items.map(item => item.insertText)
        expect(completions.items.length).toBeGreaterThan(0)
        expect(texts).toMatchInlineSnapshot(
            `
          [
            "return a + b;",
          ]
        `
        )
    }, 10_000)

    it('autocomplete/execute multiline (non-empty result)', async () => {
        // Open merge sort with a full algorithm implementation to add it to the context.
        // Otherwise there's a higher chance that model won't complete the code because of all
        // the functions in the workspace missing implementations.
        const uri1 = workspace.file('src', 'mergeSort.ts')
        await client.openFile(uri1)

        const uri = workspace.file('src', 'bubbleSort.ts')
        await client.openFile(uri)

        const completions = await client.request('autocomplete/execute', {
            uri: uri.toString(),
            position: { line: 1, character: 4 },
            triggerKind: 'Automatic',
        })

        const completionID = completions.items[0].id
        const texts = completions.items.map(item => item.insertText)

        client.notify('autocomplete/completionSuggested', { completionID })
        // Wait for the completion visibility timeout we use to ensure users read a completion.
        await client.request('testing/autocomplete/awaitPendingVisibilityTimeout', null)
        client.notify('autocomplete/completionAccepted', { completionID })

        const completionEvent = await client.request('testing/autocomplete/completionEvent', {
            completionID,
        })

        expect(completionEvent?.read).toBe(true)
        expect(completionEvent?.acceptedAt).toBeTruthy()

        expect(completions.items.length).toBeGreaterThan(0)
        expect(texts).toMatchSnapshot()
    }, 10_000)
})
