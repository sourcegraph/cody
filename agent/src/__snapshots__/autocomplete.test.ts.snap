// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Autocomplete > autocomplete/execute multiline (non-empty result) 1`] = `
[
  "for (let i = 0; i < arr.length; i++) {
        for (let j = 0; j < arr.length - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                let temp = arr[j]
                arr[j] = arr[j + 1]
                arr[j + 1] = temp
            }
        }
    }
    return arr",
]
`;
